page_group.php

```diff
@@ -0,0 +1,145 @@
+ * @author    Jason den Dulk <jasondendulk@catalyst-au.net>
+ * @copyright 2022, Catalyst IT
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+use core\chart_line;
+use core\chart_series;
+use tool_excimer\helper;
+use tool_excimer\output\tabs;
+use tool_excimer\page_group;
+
+require_once('../../../config.php');
+require_once($CFG->libdir.'/adminlib.php');
+
+$expiry = get_config('tool_excimer', 'expiry_fuzzy_counts');
+$pagegroupid = required_param('id', PARAM_INT);
+$monthstodisplay = optional_param('monthstodisplay', $expiry, PARAM_INT);
+
+$url = new \moodle_url('/admin/tool/excimer/page_group.php');
+$context = context_system::instance();
+
+$pagegroup = new page_group($pagegroupid);
+
+$PAGE->set_context($context);
+$PAGE->set_url($url);
+$PAGE->navbar->add($pagegroup->get('name'));
+
+admin_externalpage_setup('tool_excimer_report_page_groups');
+
+$output = $PAGE->get_renderer('tool_excimer');
+
+$data = $pagegroup->to_record();
+
+// Data for table.
+$data->month = userdate(strtotime($data->month . '01'), '%b %Y');
+$data->approxcount = pow(2, $data->fuzzycount - 1) . ' - ' . pow(2, $data->fuzzycount);
+$data->approxduration = pow(2, $data->fuzzydurationsum);
+$data->histogram = helper::make_histogram($data);
+
+// Data for charts.
+// Each duration range forms its own line on the chart.
+
+$firstmonth = (int) userdate(strtotime("$monthstodisplay months ago"), '%Y%m'); // YYYYMM format.
+
+$history = $DB->get_records_select(
+    page_group::TABLE,
+    "name = ? and month >= ?",
+    [$data->name, $firstmonth],
+    '',
+    'month, id, fuzzydurationcounts'
+);
+
+$highest = 0;
+
+// Each of these is an array holding counts for each month for a particular duration range.
+$durationseries = [];
+
+$histograms = [];
+// We make a list of histograms, while finding the highest duration range that a histogram has a count for.
+foreach ($history as $month => $record) {
+    $histograms[$month] = helper::make_histogram($record);
+    $highest = max($highest, count($histograms[$month]));
+}
+
+// Labels for the axis.
+$labels = [];
+
+// Label for each line.
+$linelabels = [];
+
+$month = $firstmonth;
+// We count through the months to get them in the right order, and to capture months that have no record in the database.
+for ($i = 0; $i < $monthstodisplay; ++$i) {
+    $labels[] = userdate(strtotime($month . '01'), '%b %Y');
+    // For each histogram, the duration counts are stored in a simple array. We count from 0 to highest here to
+    // ensure that we have the same number of entries for each month.
+    for ($rangehigh = 0; $rangehigh < $highest; ++$rangehigh) {
+        if (isset($histograms[$month][$rangehigh])) {
+            $durationseries[$rangehigh][] = $histograms[$month][$rangehigh]['value'];
+            $linelabels[$rangehigh] = get_string('fuzzydurationrange', 'tool_excimer', $histograms[$month][$rangehigh]);
+        } else {
+            $durationseries[$rangehigh][] = 0;
+        }
```

Comment:
Small suggestion. This change makes it so you're only pushing values into the array in 2 possible locations instead of 3, and removing the else statement.
```suggestion
        $rangehighvalue = 0;
        if (isset($histograms[$month][$rangehigh])) {
            $rangehighvalue = $histograms[$month][$rangehigh]['value'];
            $linelabels[$rangehigh] = get_string('fuzzydurationrange', 'tool_excimer', $histograms[$month][$rangehigh]);
        }
        $durationseries[$rangehigh][] = $rangehighvalue;
```

Reply:
done